system/security/securesocket/securesocket.cpp

```diff
@@ -1157,10 +1159,119 @@ const char* strtok__(const char* s, const char* d, StringBuffer& tok)
+    if (!rsa)
+        return false;
+
+    if (!SSL_CTX_use_RSAPrivateKey(ctx, rsa))
+        return false;
+
+    return true;
+}
+
+static bool useCertificateChainPEMBuffer(SSL_CTX *ctx, const char *certBuf, int certLen=-1)
+{
+    // this routine based on code originally from:
+    // https://stackoverflow.com/questions/3810058/read-certificate-files-from-memory-instead-of-a-file-using-openssl
+
+    // can have multiple certs in buffer
+
+    OwnedEVPBio cbio(BIO_new_mem_buf(certBuf, certLen));
+    if (!cbio)
+        return false;
+
+    OwnedX509StkPtr infoStk(PEM_X509_INFO_read_bio(cbio, NULL, NULL, NULL));
+    if (!infoStk)
+        return false;
+
+    bool first = true;
+    X509_INFO *infoVal;
+    for (int i=0; i<sk_X509_INFO_num(infoStk); i++)
+    {
+        infoVal = sk_X509_INFO_value(infoStk, i);
+        if (infoVal->x509)
+        {
+            // First cert is server/main cert. Remaining, if any, are intermediate certs.
+            if (first)
+            {
+                first = false;
+
+                // Set server certificate. Note that this operation increments the
+                // reference count, which means that it is okay for cleanup to free it.
+                if (!SSL_CTX_use_certificate(ctx, infoVal->x509))
+                    return false;
+
+                if (ERR_peek_last_error() != 0)
+                    return false;
+
+                // Get ready to store intermediate certs, if any.
+                SSL_CTX_clear_chain_certs(ctx);
+            }
+            else
+            {
+                // Add intermediate cert to chain.
+                if (!SSL_CTX_add0_chain_cert(ctx, infoVal->x509))
+                    return false;
+
+                // Above function doesn't increment cert reference count. NULL the
+                // reference to it in order to prevent it from being freed during cleanup.
+                infoVal->x509 = NULL;
+            }
+        }
+    }
+
+    return true;
+}
+
+static bool setVerifyCertsPEMBuffer(SSL_CTX *ctx, const char *caCertBuf, int caCertLen=-1)
+{
+    // this routine based on code originally from:
+    // https://stackoverflow.com/questions/5052563/c-openssl-use-root-ca-from-buffer-rather-than-file-ssl-ctx-load-verify-locat
+
+    // can have multiple certs in buffer
+
+    OwnedEVPBio cbio(BIO_new_mem_buf(caCertBuf, caCertLen));
+    if (!cbio)
+        return false;
+
+    OwnedX509Store store(X509_STORE_new());
+    if (!store)
+        return false;
+
+    OwnedX509StkPtr infoStk(PEM_X509_INFO_read_bio(cbio, NULL, NULL, NULL));
+    if (!infoStk)
+        return false;
+
+    X509_INFO *infoVal;
+    for (int i = 0; i < sk_X509_INFO_num(infoStk); i++)
+    {
+        infoVal = sk_X509_INFO_value(infoStk, i);
+        if (infoVal->x509)
+        {
+            if (!X509_STORE_add_cert(store, infoVal->x509))
+                return false;
+
+            infoVal->x509 = NULL;
+        }
+    }
+
+    SSL_CTX_set_cert_store(ctx, store);
```

Comment:
What happens when the OwnedX509Store above goes out of scope but the pointer has been passed to SSL_CTX_set_cert_store?

Should you do:
    SSL_CTX_set_cert_store(ctx, store.getClear());

instead?

Reply:
Yes, thank you.  Need to free previous.

Reply:
But then does it get freed when ref count goes to zero ?  Or it is already zero at this point ?

Reply:
When would it ever get freed ?  I am thinking we need to free it when we free ctx ?

Reply:
Maybe a store should be added to CSecureSocketContext ?

Reply:
@mckellyln I think this is the difference between SSL_CTX_set_cert_store and SSL_CTX_set1_cert_store.  One takes ownership and one increments and decrements the reference count.  Either case needs testing to make sure there is no leak or extra free.

I think the only reason to add an m_store is if the store is shared by multiple contexts otherwise it's extra complexity with nothing to gain.